compiler/rustc_interface/src/passes.rs

```diff
@@ -219,9 +219,13 @@ pub fn register_plugins<'a>(
     let crate_types = util::collect_crate_types(sess, &krate.attrs);
     sess.init_crate_types(crate_types);
 
-    let disambiguator = util::compute_crate_disambiguator(sess);
-    sess.crate_disambiguator.set(disambiguator).expect("not yet initialized");
-    rustc_incremental::prepare_session_directory(sess, &crate_name, disambiguator)?;
+    let stable_crate_id = StableCrateId::new(
+        crate_name,
+        sess.crate_types().contains(&CrateType::Executable),
+        sess.opts.cg.metadata.clone(),
+    );
+    sess.stable_crate_id.set(stable_crate_id).expect("not yet initialized");
```

Comment:
I'd love if we could compute that even earlier so that `stable_crate_id` does not need to be mutable at all.

Reply:
That is not possible until https://github.com/rust-lang/rust/pull/83744 has landed and until the future incompat lint has been changed into a hard error. The `Session` is created before parsing and expanding the source code. Both the crate name and crate type can currently only be determined after expansion.

Reply:
#83744 looks great. I hope it gets merged `:)`